pelican/tests/support.py

```diff
@@ -236,24 +212,103 @@ def diff_subproc(first, second):
     )
 
 
-class LoggedTestCase(unittest.TestCase):
-    """A test case that captures log messages."""
+class LogCountHandler(BufferingHandler):
+    """Capturing and counting logged messages."""
 
-    def setUp(self):
-        super().setUp()
-        self._logcount_handler = LogCountHandler()
-        logging.getLogger().addHandler(self._logcount_handler)
+    def __init__(self, capacity=1000):
+        super().__init__(capacity)
 
-    def tearDown(self):
-        logging.getLogger().removeHandler(self._logcount_handler)
-        super().tearDown()
-
-    def assertLogCountEqual(self, count=None, msg=None, **kwargs):
-        actual = self._logcount_handler.count_logs(msg=msg, **kwargs)
-        self.assertEqual(
-            actual, count,
-            msg='expected {} occurrences of {!r}, but found {}'.format(
-                count, msg, actual))
+    @classmethod
+    @contextmanager
+    def examine(cls, loggerObj):
+        """
+        Context in which a logger's propagated messages can be examined.
+
+        Yields
+        ======
+        A handle to ``LogCountHandler.assert_count`` that has been added to the
+        specified logger for the duration of the context.
+
+        The yielded caller can be used to assert whether a certain number of
+        log messages have occurred within the context.
+        """
+        hnd = cls()
+        try:
+            loggerObj.addHandler(hnd)
+            yield hnd.assert_count
+        finally:
+            loggerObj.removeHandler(hnd)
+
+    def assert_count(
+        self,
+        count: int,
+        msg: Optional[str] = None,
+        level: Optional[int] = None,
+        as_regex: bool = False
+    ):
+        """
+        Assert how often the specified messages have been handled.
+
+        Raises
+        -------
+        AssertionError
+        """
+        occurances = self.count_logs(msg, level, as_regex)
+        if count != occurances:
+            report = 'Logged occurrence'
+            if msg is not None:
+                report += ' of {!r}'.format(msg)
+
+            if level is not None:
+                raise AssertionError(
+                    ' at {}'.format(logging.getLevelName(level))
+                )
+            raise AssertionError(
+                report + ': expected/found {}/{}'.format(count, occurances)
+            )
+
+    def match_record(
+        self,
+        pattern: re.Pattern,
+        record: logging.LogRecord,
+        level: Optional[int]
+    ) -> Optional[re.Match]:
+        """
+        Return regex object if pattern found in message at specified severity.
+        """
+        if level is not None and level != record.levelno:
+            return None
+
```

Comment:
I think the reason I did this with regex objects instead of direct strings is because of the possible need to escape characters when a regex isn't actually desired (see ``count_logs()`` def below) and wanting to leave that decision up to the caller.

Also -- now that I'm looking at this again, I hate the comment I put 2 lines down: it sounds like it describes what's happening -- it's not!  It's *actually* a "how-to" (which I should have put in the docstring -_-) ...